### PR TITLE
The experimental decoder to truncate on out-of-bounds boolean input.

### PR DESCRIPTION
**Pull request by axic**
Part of #5844.

### CODE REVIEW

File: libsolidity/codegen/ABIFunctions.cpp [hunk 1]
```diff
@@ -212,8 +212,15 @@ string ABIFunctions::cleanupFunction(Type const& _type, bool _revertOnFailure)
 			templ("body", "cleaned := value");
 			break;
 		case Type::Category::Bool:
-			templ("body", "cleaned := iszero(iszero(value))");
+		{
+			Whiskers w("if gt(value, 1)) { <failure> } cleaned := value");
```
Comment (axic) [inline libsolidity/codegen/ABIFunctions.cpp]: How did this succeed with the tests? It has two closing parentheses.

**Comment by chriseth** [comment]
Can you do it for all types in one go? Otherwise we might miss some cases.

**Comment by axic** [comment]
> Can you do it for all types in one go? Otherwise we might miss some cases.

@chriseth If I recall everything correctly: `cleanup` is used for incoming data, `convert` is used for data to be encoded. The goal was to be strict on incoming data and enforce zero padding for short types.

That means apart from `Bool`, `FixedBytes` and the `Integer` case must be fixed. That should trigger the rest, like `Address`.

**Comment by axic** [comment]
I think we need to clarify this before we make the ABIEncoderV2 non-experimental. Ideally this would go in this or next release, iff we decide to go this way.

**Comment by axic** [comment]
It fails on this one currently:
```
BOOST_AUTO_TEST_CASE(value_types)
{
        string sourceCode = R"(
                contract C {
                        event E(uint a, uint16 b, uint24 c, int24 d, bytes3 x, bool, C);
                        function f() public {
                                bytes6 x = hex"1bababababa2";
                                bool b;
                                assembly { b := 7 }
                                C c;
                                assembly { c := sub(0, 5) }
                                emit E(10, uint16(uint256(-2)), uint24(0x12121212), int24(int256(-1)), bytes3(x), b, c);
                        }
                }
        )";
```

It means that `cleanup` is used on the emit (`conversion` calls it internally) where `bool` has the underlying value of 7. There's no compilation error, but a runtime error.

Technically we can fix that so cleanup won't insert the runtime check, but that requires a new flag on `cleanup`. Are we ok with that?

**Comment by chriseth** [comment]
For encoding integers, I think we should not assert, because they are allowed to have dirty higher bits on intermediate values (and truncating them is a defined operation). It is similar with fixed bytes.

I think bool could be treated the same way as a two-valued enum.

**Comment by axic** [comment]
> I think bool could be treated the same way as a two-valued enum.

That is what happens in the PR.

**Comment by axic** [comment]
Also I'd say it is slightly inconsistent because for enum it generates the same runtime check on encoding, which is unneeded - would just need the mask.

**Comment by axic** [comment]
> For encoding integers, I think we should not assert, because they are allowed to have dirty higher bits on intermediate values (and truncating them is a defined operation). It is similar with fixed bytes.

So this PR was originally meant to handle *incoming* data and not internal data.

**Comment by axic** [comment]
I've looked into a bit more and first hunch was to introduce an enum:
```
+       enum class CleanupMode {
+               Silent,
+               RevertOnFailure,
+               AbortOnFailure
+       };
+
-       std::string cleanupFunction(Type const& _type, bool _revertOnFailure = false);
+       std::string cleanupFunction(Type const& _type, CleanupMode _cleanupMode = CleanupMode::AbortOnFailure);
```

But I actually think we should split `cleanup` into two: `cleanup` and `validate`.

Another side note: some "cleanups" recurse, but the `_revertOnFailure` flag is not passed down.

**Comment by chriseth** [comment]
I would say there are three options for each cleanup:
 - wrap / clean 
 - validate and assert
 - validate and revert

Does it make sense to provide both clean and validate for the same type?

**Comment by axic** [comment]
In my opinion:
- Cleaning is used for encoding. Internally we can have dirty bits, but we do not want to encode that.
- Validation is used for decoding. For external input we do not want dirty bits.

I've tried an option too see how much change is it to split `cleanupFunction` into `cleanupFunction` and `validatorFunction` and it is not too much - but is quite a bit of code duplication, **before** we add actual validation which is missing.

**Comment by axic** [comment]
Handling this in #5815.

**Pull request closed without merging**